README.md
# Factom Node Monitoring and Alert Tool

Factom-monitord is a tool to monitor the health of one or more factomd nodes. It will alert the user via a text message or phone call in the event that a node crashes or stalls.

Factom-monitord allows users to implement a personnel rota; only the person on call at the time of a crash or stall event will be notified. Moreover, the rota allows users to set a priority rating for each person in the rota, which factom-monitord uses to select a person to alert when there is a gap in the rota.

Users must have an account with [Twilio](https://www.twilio.com/) to operate this script.

## Installing NodeJS

First, install NodeJS on your system.

### Ubuntu

```
curl -sL https://deb.nodesource.com/setup_10.x | sudo -E bash -
sudo apt-get install -y nodejs
```

### MacOS

Visit the NodeJS website to download the installer package: https://nodejs.org/en/#download

Alternatively, if you have Homebrew installed:

```
brew install node
```

## Installing the Monitoring Tool

Clone the repo into your target directory. Then, navigate into the project folder and type:

```
npm install
```

## Installing PM2 (optional but recommended)

PM2 is a process manager built for NodeJS. It is a convenient and easy way to manage the script. To install, type:

```
npm install pm2 -g
```

(Note: You need to run the above command with root permissions)

## Running

First, copy config.yaml.copy to config.yaml.

Then, fill out all of the details in config.yaml. It is important that you read the comments in the config file carefully otherwise the script may not work as expected.

Second, make sure the firewall for the control panel port is open to factom-monitord's public IP on each target host.

### Without PM2

To run the script without PM2, make sure you're in the project's root directory then run:

```
node factom-monitord.js
```

### With PM2

With PM2, make sure you are in the project's root directory then run:

```
pm2 start factom-monitord.js
```

Next, save the current process list:

```
pm2 save
```

Finally, to make sure the script starts on boot, run:

```
pm2 startup
```

then follow the onscreen instructions.

## Built With

-   [Twilio](https://www.twilio.com/)

## Authors

-   **Alex Carrithers**

## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE) file for details


changelog
changelog v0.4.0

- New: introduce rota system
- Update: change name to factom-monitord
- Bug fix: give user feedback and trigger alert if fail to make initial connection

changelog v0.3.0

- Complete rewrite
- Add support for monitoring multiple nodes
- Add support for text message alerts
- Polls node more frequently but also less vulnerable to temp connection outages
- Change to YAML config file


test/rota.test.js
const assert = require('assert');
const moment = require('moment');

class TestRota {
    constructor(config, today, time) {
        this.config = config;
        this.today = today;
        this.time = time;
        this.voice = config.node.voice,
        this.text = config.node.text,
        this.alertOffDutyByText = config.alertOffDutyByText;
        this.textRecipients = [];
        this.onDuty = {};
        this.onDuty.text = null;
        this.onDuty.voice = null;
    }

    alertUser() {
        let personOnDuty = this.getPersonOnDuty();

        if (this.voice === true) {
            this.onDuty.voice = true;
        }

        if (this.text === true) {
            this.onDuty.text = true;
        }

        if (this.alertOffDutyByText === true) {
            this.config.rota.forEach(person => {
                if (person.name !== personOnDuty.name) {
                    this.textRecipients.push(person);
                }
            });
        }
    }

    getPersonOnDuty() {
        const peopleOnDuty = [];
        const today = moment(this.today, 'dddd').format('dddd');
        const time = moment(this.time, 'HH:mm').format('HH:mm');
        this.config.rota.forEach(person => {
            const onDutyToday = person.onDuty[`${today}`];
            if (onDutyToday) {
                const startTime = moment(onDutyToday.split(' - ')[0], 'HH:mm').format('HH:mm');
                const endTime = moment(onDutyToday.split(' - ')[1], 'HH:mm').format('HH:mm');
                if (startTime <= time && endTime >= time) {
                    peopleOnDuty.push(person);
                }
            }
        });

        if (peopleOnDuty.length > 0) {
            return this.determinePriority(peopleOnDuty);
        } else {
            return this.determinePriority(this.config.rota);
        }
    }

    determinePriority(contactArray) {
        return contactArray.reduce((highestPriority, currentPerson) => {
            if (currentPerson.priority > highestPriority.priority) {
                return highestPriority = currentPerson;
            }
            return highestPriority;
        });
    }


}

describe('Check that the script correctly parses the rota', () => {
    const config = {
        rota: [{
            name: 'Barry',
            phoneNumber: '447845609873',
            onDuty: {
                Monday: '16:00 - 23:59',
                Tuesday: '16:00 - 23:59',
                Wednesday: '16:00 - 23:59',
                Thursday: '16:00 - 23:59',
                Saturday: '16:00 - 23:59',
            },
            priority: '3',
        },
        {
            name: 'Bob',
            phoneNumber: '447845609873',
            onDuty: {
                Monday: '00:00 - 08:00',
                Tuesday: '00:00 - 08:00',
                Wednesday: '00:00 - 08:00',
                Thursday: '00:00 - 08:00',
                Friday: '00:00 - 08:00',
                Sunday: '16:00 - 23:59',
            },
            priority: '1',
        },
        {
            name: 'Bill',
            phoneNumber: '447845609873',
            onDuty: {
                Monday: '08:00 - 16:00',
                Tuesday: '08:00 - 16:00',
                Wednesday: '08:00 - 16:00',
                Thursday: '08:00 - 16:00',
                Saturday: '08:00 - 16:00',
                Sunday: '16:00 - 23:59',
            },
            priority: '2',
        },
        {
            name: 'Ben',
            phoneNumber: '447845609873',
            onDuty: {
                Monday: '16:00 - 23:59',
                Tuesday: '16:00 - 23:59',
                Wednesday: '16:00 - 23:59',
                Thursday: '16:00 - 23:59',
                Saturday: '16:00 - 23:59',
            },
            priority: '3',
        }],
        node: {
            voice: null,
            text: null
        },
    };

    it('Should select the person currently on duty', () => {
        const testRota = new TestRota(config, 'Monday', '08:00');
        const personOnDuty = testRota.getPersonOnDuty();

        assert(personOnDuty.name === 'Bill');
    });

    it('Should select the person with the highest priority when no one is on duty', () => {
        const testRota = new TestRota(config, 'Sunday', '14:00');
        const personOnDuty = testRota.getPersonOnDuty();

        assert(personOnDuty.name === 'Ben' || personOnDuty.name === 'Barry');
    });

    it('Should select only one person when two have equal priority and on duty together', () => {
        const testRota = new TestRota(config, 'Saturday', '17:00');
        const personOnDuty = testRota.getPersonOnDuty();

        assert(personOnDuty.name === 'Ben' || personOnDuty.name === 'Barry');
    });

    it('Should select the right person when two are on duty and one has higher higher priority', () => {
        const testRota = new TestRota(config, 'Sunday', '17:00');
        const personOnDuty = testRota.getPersonOnDuty();

        assert(personOnDuty.name === 'Bill');
    });

    it('Should only call the person on duty', () => {
        config.alertOffDutyByText = false;
        config.node.voice = true;
        config.node.text = false;
        const testRota = new TestRota(config, 'Sunday', '17:00');
        testRota.alertUser();

        assert(testRota.onDuty.voice === true);
        assert(testRota.onDuty.text === null);
        assert(testRota.textRecipients.length === 0);
    });

    it('Should only text the person on duty', () => {
        config.alertOffDutyByText = false;
        config.node.voice = false;
        config.node.text = true;
        const testRota = new TestRota(config, 'Sunday', '17:00');
        testRota.alertUser();

        assert(testRota.onDuty.voice === null);
        assert(testRota.onDuty.text === true);
        assert(testRota.textRecipients.length === 0);
    });

    it('Should call the person on duty and alert everyone off duty by text', () => {
        config.alertOffDutyByText = true;
        config.node.voice = true;
        config.node.text = true;
        const testRota = new TestRota(config, 'Sunday', '17:00');
        testRota.alertUser();

        assert(testRota.onDuty.voice === true);
        assert(testRota.onDuty.text === true);
        assert(testRota.textRecipients.length === 3);
    });
});


test/failure.test.js
const assert = require('assert');

describe('Check if the script handles failure logic properly', () => {
    function Node(node) {
        this.status = node.status;
    }

    it('should result in a stalled status and an alert call', async () => {
        Node.prototype.monitor = function() {
            this.lastBlockTime = Date.now() - 1800001;
        };
        const node = new Node({ status: 'online' });
        await node.monitor();

        if (Date.now() - node.lastBlockTime > 1800000 && node.status === 'online') {
            node.status = 'stalled';
            node.alertTriggered = true;
        }
        assert(node.status === 'stalled');
        assert(node.alertTriggered === true);
    });

    it('should result in an offline status and an alert call', async () => {
        Node.prototype.checkHeight = function() {
            return null;
        };
        const node = new Node({ status: 'online' });
        const nodeHeight = node.checkHeight();

        if (!nodeHeight && node.status !== 'offline') {
            node.status = 'offline';
            node.alertTriggered = true;
        }
        assert(node.status === 'offline');
        assert(node.alertTriggered === true);
    });
});


tests/failure.test.js
const assert = require('assert');
const axios = require('axios');

describe('Check if the script handles failure logic properly', () => {
    function Node(node) {
        this.status = node.status;
    }

    it('should result in a stalled status and an alert call', async () => {
        Node.prototype.monitor = function() {
            this.lastBlockTime = Date.now() - 1800001;
        };
        const node = new Node({ status: 'online' });
        await node.monitor();

        if (Date.now() - node.lastBlockTime > 1800000 && node.status === 'online') {
            node.status = 'stalled';
            node.alertTriggered = true;
        }
        assert(node.status === 'stalled');
        assert(node.alertTriggered === true);
    });

    it('should result in an offline status and an alert call', async () => {
        Node.prototype.checkHeight = function() {
            return null;
        };
        const node = new Node({ status: 'online' });
        const nodeHeight = node.checkHeight();

        if (!nodeHeight && node.status !== 'offline') {
            node.status = 'offline';
            node.alertTriggered = true;
        }
        assert(node.status === 'offline');
        assert(node.alertTriggered === true);
    });
});


set-config.js
const fs = require('fs')

const inquirer = require('inquirer')

var questions = [
    {
        type: 'input',
        name: 'SID',
        message: 'What is your Twilio account SID?'
    },
    {
        type: 'input',
        name: 'token',
        message: 'What is your Twilio auth token?'
    },
    {
        type: 'input',
        name: 'from',
        message: 'What phone number does your alert call come from?'
    },
    {
        type: 'input',
        name: 'to',
        message: 'What phone number do you receive your alert call to?'
    },
    {
        type: 'input',
        name: 'voiceUrl',
        message: 'What is your Twilio voice URL?'
    },
    {
        type: 'input',
        name: 'host',
        message: 'What is the IP address of the host you are monitoring?',
        default: 'localhost:8090'
    }
]


inquirer.prompt(questions).then(answers => {
    fs.writeFileSync('./config.json', JSON.stringify(answers))
    console.log('You can now start the monitor using `node monitor.js`')
})


factom-monitord.js
const fs = require('fs');

const moment = require('moment');
const axios = require('axios');

const yaml = require('yaml-js');
const config = yaml.load(fs.readFileSync('./config.yaml'));

const Twilio = require('twilio');
const client = new Twilio(config.twilio.SID, config.twilio.authToken);

class Node {
    constructor(node) {
        this.ip = node.ip;
        this.port = node.port;
        this.name = node.name;
        this.voice = node.voice;
        this.text = node.text;
        this.height = 0;
        this.lastBlockTime = 0;
    }

    // prettier-ignore
    async evaluateNodeState() {
        const nodeHeight = await this.checkHeight();
        
        // success case
        if (nodeHeight > this.height || (nodeHeight && this.status === 'offline')) {
            this.height = nodeHeight;
            this.lastBlockTime = Date.now();
            this.status = 'online';
            console.log('\x1b[32m', `${this.name}:`, '\x1b[0m', `blockheight - ${this.height}`);
        // crash case
        } else if (!nodeHeight && this.status === 'online') {
            this.status = 'offline';
            console.log('\x1b[31m', `${this.name}:`, '\x1b[0m', `node is offline - ${this.height}`);
            this.alertPersonOnDuty();
        // stall case
        } else if (Date.now() - this.lastBlockTime > 1200000 && this.status === 'online') {
            this.status = 'stalled';
            console.log('\x1b[31m', `${this.name}:`, '\x1b[0m', `node has stalled - ${this.height}`);
            this.alertPersonOnDuty();
        // initial contact failure case
        } else if (!nodeHeight && this.height === 0 && this.status !== 'offline') {
            this.status = 'offline';
            console.log('\x1b[31m', `${this.name}:`, '\x1b[0m', `cannot make initial contact. Are you sure factomd is running and the port is open?`);
            this.alertPersonOnDuty();
        }
    }

    // prettier-ignore
    async checkHeight() {
        for (let i = 0; i < 10; i++) {
            try {
                // had to add timeout, otherwise a failed node could take an absurdly long time to trigger an alert
                const controlPanelHeight = await axios.get(`http://${this.ip}:${this.port}/factomdBatch?batch=myHeight`, {
                    timeout: 2500
                });
                return controlPanelHeight.data[0].Height;
            } catch(err) {
                // Math.pow() implements incremental back-off for node retries
                await new Promise(resolve => setTimeout(resolve, Math.pow(3, i)));
            }
        }
        return null;
    }

    alertPersonOnDuty() {
        const { name, phoneNumber } = this.getPersonOnDuty();

        console.log(`Alerting ${name} on ${phoneNumber}`);

        if (this.voice === true) {
            client.api.calls
                .create({
                    url: config.twilio.voiceUrl,
                    to: phoneNumber,
                    from: config.twilio.from
                })
                .done();
        }

        if (this.text === true) {
            client.messages
                .create({
                    body: `${this.name} is ${this.status}.`,
                    to: phoneNumber,
                    from: config.twilio.from
                })
                .done();
        }
    }

    // prettier-ignore
    getPersonOnDuty() {
        const peopleOnDuty = [];
        const now = moment().format('HH:mm');
        const today = moment().format('dddd');
        config.rota.forEach(person => {
            const personOnDutyToday = person.onDuty[today] || person.onDuty[today.toLowerCase()];
            if (personOnDutyToday) {
                const shiftStart = personOnDutyToday[0];
                const shiftEnd = personOnDutyToday[1];
                if (shiftStart <= now && shiftEnd >= now) {
                    peopleOnDuty.push(person);
                }
            }
        });

        if (peopleOnDuty.length > 0) {
            return this.determineAlertPriority(peopleOnDuty);
        } else {
            return this.determineAlertPriority(config.rota);
        }
    }

    determineAlertPriority(contactArray) {
        return contactArray.reduce((highestPerson, currentPerson) => {
            if (currentPerson.priority > highestPerson.priority) {
                return currentPerson;
            }
            return highestPerson;
        });
    }
}

async function startMonitoring() {
    config.nodes.forEach(async nodeObject => {
        const node = new Node(nodeObject);

        while (true) {
            await node.evaluateNodeState();
            await new Promise(resolve => setTimeout(resolve, 30000));
        }
    });
}

startMonitoring();


factom-node-monitord.js
const fs = require('fs');

const moment = require('moment');
const axios = require('axios');

const yaml = require('yaml-js');
const config = yaml.load(fs.readFileSync('./config.yaml'));

const Twilio = require('twilio');
const client = new Twilio(config.twilio.SID, config.twilio.authToken);

class Node {
    constructor(node) {
        this.ip = node.ip;
        this.port = node.port;
        this.name = node.name;
        this.voice = node.voice;
        this.text = node.text;
        this.height = 0;
        this.lastBlockTime = 0;
    }

    // prettier-ignore
    async evaluateNodeState() {
        const nodeHeight = await this.checkHeight();
        
        // success case
        if (nodeHeight > this.height || (nodeHeight && this.status === 'offline')) {
            this.height = nodeHeight;
            this.lastBlockTime = Date.now();
            this.status = 'online';
            console.log('\x1b[32m', `${this.name}:`, '\x1b[0m', `blockheight - ${this.height}`);
        // crash case
        } else if (!nodeHeight && this.status === 'online') {
            this.status = 'offline';
            console.log('\x1b[31m', `${this.name}:`, '\x1b[0m', `node is offline - ${this.height}`);
            this.alertPersonOnDuty();
        // stall case
        } else if (Date.now() - this.lastBlockTime > 1200000 && this.status === 'online') {
            this.status = 'stalled';
            console.log('\x1b[31m', `${this.name}:`, '\x1b[0m', `node has stalled - ${this.height}`);
            this.alertPersonOnDuty();
        // initial contact failure case
        } else if (!nodeHeight && this.height === 0 && this.status !== 'offline') {
            this.status = 'offline';
            console.log('\x1b[31m', `${this.name}:`, '\x1b[0m', `cannot make initial contact. Are you sure factomd is running and the port is open?`);
            this.alertPersonOnDuty();
        }
    }

    // prettier-ignore
    async checkHeight() {
        for (let i = 0; i < 10; i++) {
            try {
                // had to add timeout, otherwise a failed node could take an absurdly long time to trigger an alert
                const controlPanelHeight = await axios.get(`http://${this.ip}:${this.port}/factomdBatch?batch=myHeight`, {
                    timeout: 3000
                });
                return controlPanelHeight.data[0].Height;
            } catch(err) {
                // Math.pow() implements incremental back-off for node retries
                await new Promise(resolve => setTimeout(resolve, Math.pow(3, i)));
            }
        }
        return null;
    }

    alertPersonOnDuty() {
        const { name, phoneNumber } = this.getPersonOnDuty();

        console.log(`Alerting ${name} on ${phoneNumber}`);

        if (this.voice === true) {
            client.api.calls
                .create({
                    url: config.twilio.voiceUrl,
                    to: phoneNumber,
                    from: config.twilio.from
                })
                .done();
        }

        if (this.text === true) {
            client.messages
                .create({
                    body: `${this.name} is ${this.status}.`,
                    to: phoneNumber,
                    from: config.twilio.from
                })
                .done();
        }
    }

    // prettier-ignore
    getPersonOnDuty() {
        const peopleOnDuty = [];
        const now = moment().format('HH:mm');
        const today = moment().format('dddd');
        config.rota.forEach(person => {
            const personOnDutyToday = person.onDuty[today] || person.onDuty[today.toLowerCase()];
            if (personOnDutyToday) {
                const shiftStart = personOnDutyToday[0];
                const shiftEnd = personOnDutyToday[1];
                if (shiftStart <= now && shiftEnd >= now) {
                    peopleOnDuty.push(person);
                }
            }
        });

        if (peopleOnDuty.length > 0) {
            return this.determineAlertPriority(peopleOnDuty);
        } else {
            return this.determineAlertPriority(config.rota);
        }
    }

    determineAlertPriority(contactArray) {
        return contactArray.reduce((highestPerson, currentPerson) => {
            if (currentPerson.priority > highestPerson.priority) {
                return currentPerson;
            }
            return highestPerson;
        });
    }
}

async function startMonitoring() {
    config.nodes.forEach(async nodeObject => {
        const node = new Node(nodeObject);

        while (true) {
            await node.evaluateNodeState();
            await new Promise(resolve => setTimeout(resolve, 30000));
        }
    });
}

startMonitoring();


monitor.js
const fs = require('fs');

const moment = require('moment');
const axios = require('axios');

const yaml = require('yaml-js');
const config = yaml.load(fs.readFileSync('./config.yaml'));

const Twilio = require('twilio');
const client = new Twilio(config.twilio.SID, config.twilio.authToken);

class Node {
    constructor(node) {
        this.ip = node.ip;
        this.port = node.port;
        this.name = node.name;
        this.voice = node.voice;
        this.text = node.text;
        this.height = 0;
        this.lastBlockTime = 0;
    }

    // prettier-ignore
    async evaluateNodeState() {
        const nodeHeight = await this.checkHeight();

        if (!nodeHeight && this.status === 'online') {
            this.status = 'offline';
            console.log('\x1b[31m', `${this.name}:`, '\x1b[0m', `node is offline - ${this.height}`);
            this.alertPersonOnDuty();
        } else if (Date.now() - this.lastBlockTime > 1200000 && this.status === 'online') {
            this.status = 'stalled';
            console.log('\x1b[31m', `${this.name}:`, '\x1b[0m', `node has stalled - ${this.height}`);
            this.alertPersonOnDuty();
        } else if (nodeHeight > this.height || (nodeHeight && this.status === 'offline')) {
            this.height = nodeHeight;
            this.lastBlockTime = Date.now();
            this.status = 'online';
            console.log('\x1b[32m', `${this.name}:`, '\x1b[0m', `blockheight - ${this.height}`);
        }
    }

    // prettier-ignore
    async checkHeight() {
        for (let i = 0; i < 10; i++) {
            try {
                const controlPanelHeight = await axios.get(`http://${this.ip}:${this.port}/factomdBatch?batch=myHeight`);
                return controlPanelHeight.data[0].Height;
            } catch(err) {
                await new Promise(resolve => setTimeout(resolve, 3000));
            }
        }
        return null;
    }

    alertPersonOnDuty() {
        const { name, phoneNumber } = this.getPersonOnDuty();

        console.log(`Alerting ${name} at ${phoneNumber}`);

        if (this.voice === true) {
            client.api.calls
                .create({
                    url: config.twilio.voiceUrl,
                    to: phoneNumber,
                    from: config.twilio.from
                })
                .done();
        }

        if (this.text === true) {
            client.messages
                .create({
                    body: `${this.name} is ${this.status}.`,
                    to: phoneNumber,
                    from: config.twilio.from
                })
                .done();
        }
    }

    // prettier-ignore
    getPersonOnDuty() {
        const peopleOnDuty = [];
        const today = moment().format('dddd');
        const time = moment().format('HH:mm');
        config.rota.forEach(person => {
            const onDutyToday = person.onDuty[today];
            if (onDutyToday) {
                const startTime = moment(onDutyToday.split(' - ')[0], 'HH:mm').format('HH:mm');
                const endTime = moment(onDutyToday.split(' - ')[1], 'HH:mm').format('HH:mm');
                if (startTime <= time && endTime >= time) {
                    peopleOnDuty.push(person);
                }
            }
        });

        if (peopleOnDuty.length > 0) {
            return this.determineAlertPriority(peopleOnDuty);
        } else {
            return this.determineAlertPriority(config.rota);
        }
    }

    determineAlertPriority(contactArray) {
        return contactArray.reduce((highestPerson, currentPerson) => {
            if (currentPerson.priority > highestPerson.priority) {
                return currentPerson;
            }
            return highestPerson;
        });
    }
}

async function startMonitoring() {
    config.nodes.forEach(async nodeObject => {
        const node = new Node(nodeObject);

        while (node) {
            await node.evaluateNodeState();
            await new Promise(resolve => setTimeout(resolve, 30000));
        }
    });
}

startMonitoring();
